### PR TITLE
doc: workaround to npm shrinkwrap development dependencies bug

### PR DESCRIPTION
**Pull request by jviotti**
There are some npm versions that for some strange reason will attempt to
put *every* single development dependency in the `npm-shinkwrap.json`
file when editing the dependencies list.

This commit describes a workaround I've found to ensure only production
dependencies are added to the file.

I also renamed our `npm run shrinkwrap` script to `npm run
clean-shrinkwrap` since in some recent versions of npm, running `npm
shrinkwrap` will trigger our custom script instead of the original
implementation.

Signed-off-by: Juan Cruz Viotti <jviotti@openmailbox.org>

### CODE REVIEW

**Comment by lurch** [comment]
Just out of curiosity I tried doing this myself, and the first two steps work, but I get an error on the last step:
```bash
$ npm shrinkwrap

> etcher@1.0.0-beta.18 shrinkwrap /home/andrew/github/resin-io/etcher-clean
> node ./scripts/shrinkwrap.js

module.js:471
    throw err;
    ^

Error: Cannot find module 'jsonfile'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/andrew/github/resin-io/etcher-clean/scripts/shrinkwrap.js:20:18)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)

npm ERR! Linux 3.19.0-80-generic
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "shrinkwrap"
npm ERR! node v6.9.4
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! etcher@1.0.0-beta.18 shrinkwrap: `node ./scripts/shrinkwrap.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the etcher@1.0.0-beta.18 shrinkwrap script 'node ./scripts/shrinkwrap.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the etcher package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node ./scripts/shrinkwrap.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs etcher
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls etcher
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/andrew/github/resin-io/etcher-clean/npm-debug.log
```
:-/

**Comment by jviotti** [comment]
Did you run `npm shrinkwrap` or `npm run shrinkwrap`? Notice you should run the former on the workaround.

**Comment by lurch** [comment]
I typed `npm shrinkwrap`, exactly as documented in this PR.
Maybe it works for you because you have different NPM modules globally installed than I do?

**Comment by jviotti** [comment]
Oh, that's definitely strange. What npm version are you running? Maybe your version tried to run user scripts over pre-defined functions? I'm on 3.9.6.

**Comment by lurch** [comment]
The npm version is included in the debug output just a few lines up ^^  ;-)

**Comment by jviotti** [comment]
I can indeed reproduce in your version, but I couldn't find any workaround. I'll rename our custom script to something else.

**Comment by jviotti** [comment]
Done!

**Comment by lurch** [comment]
Yup, those three steps work now :-)

Curiously, doing a `git diff` shows that the `ecc-jsbn`, `jodid25519`, `jsbn` & `tweetnacl` modules in `npm-shrinkwrap.json` all gain an extra `"optional": true` attribute. Is that significant? ;)

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -44,12 +44,23 @@ Use the following steps to ensure everything goes flawlessly:
...
-- Run `npm run shrinkwrap`. This is a small script that ensures that operating
-  system specific dependencies that could get included in the previous step are
-  removed from `npm-shrinkwrap.json`.
+- Run `npm run clean-shrinkwrap`. This is a small script that ensures that
+  operating system specific dependencies that could get included in the
+  previous step are removed from `npm-shrinkwrap.json`.
 
 - Commit *both* `package.json` and `npm-shrinkwrap.json`.
 
+Some npm versions seem to contain an issue were all development dependencies
```
Comment (lurch) [inline docs/CONTRIBUTING.md]: Should these instructions be *above* the "Commit *both* `package.json` and `npm-shrinkwrap.json`" line?

Comment (jviotti) [inline docs/CONTRIBUTING.md]: You're right. Fixed!

**Comment by jviotti** [comment]
> Curiously, doing a git diff shows that the ecc-jsbn, jodid25519, jsbn & tweetnacl modules in npm-shrinkwrap.json all gain an extra "optional": true attribute. Is that significant? ;)

Let's ignore them from now. I took a look at those packages and they are not platform specific optional dependencies.

**Comment by lurch** [comment]
Minor typos in the commit message:
"some npm version that" -> "some npm versions that"
"dependency on the `npm-shinkwrap.json`" -> "dependency in the `npm-shinkwrap.json`"
"since it some" -> "since in some"

I don't mind if you can't be bothered to fix them though ;)

File: docs/CONTRIBUTING.md [hunk 2]
```diff
@@ -44,9 +44,20 @@ Use the following steps to ensure everything goes flawlessly:
...
-- Run `npm run shrinkwrap`. This is a small script that ensures that operating
-  system specific dependencies that could get included in the previous step are
-  removed from `npm-shrinkwrap.json`.
+- Run `npm run clean-shrinkwrap`. This is a small script that ensures that
+  operating system specific dependencies that could get included in the
+  previous step are removed from `npm-shrinkwrap.json`.
+
+Some npm versions seem to contain an issue were all development dependencies
+will be included in `npm-shrinkwrap.json` when attempting to modify it (e.g: by
+`npm install`, `npm uninstall`, etc). A bullet proof way to ensure only the
```
Comment (lurch) [inline docs/CONTRIBUTING.md]: If you want to, you could correct "bullet proof" to "bulletproof"

Comment (jviotti) [inline docs/CONTRIBUTING.md]: Fixed!

**Comment by jviotti** [comment]
All typos fixed, I'm happy you caught those!

**Pull request merged**